examples/bert/run_glue_finetuning.py

```diff
@@ -200,28 +198,60 @@ def preprocess_data(inputs, labels):
 
     # Read and preprocess GLUE task data.
     train_ds, test_ds, validation_ds = load_data(FLAGS.task_name)
-    train_ds = train_ds.batch(FLAGS.batch_size).map(
+
+    # batch_size is taken as 32.
+    train_ds = train_ds.batch(batch_size).map(
         preprocess_data, num_parallel_calls=tf.data.AUTOTUNE
     )
-    validation_ds = validation_ds.batch(FLAGS.batch_size).map(
+    validation_ds = validation_ds.batch(batch_size).map(
         preprocess_data, num_parallel_calls=tf.data.AUTOTUNE
     )
-    test_ds = test_ds.batch(FLAGS.batch_size).map(
+    test_ds = test_ds.batch(batch_size).map(
         preprocess_data, num_parallel_calls=tf.data.AUTOTUNE
     )
 
-    finetuning_model = BertClassificationFinetuner(
-        bert_model=model,
-        hidden_size=bert_config["hidden_size"],
-        num_classes=3 if FLAGS.task_name in ("mnli", "ax") else 2,
+    class BertHyperModel(keras_tuner.HyperModel):
+        """Creates a hypermodel to help with the search space for finetuning."""
+
+        def build(self, hp):
+            finetuning_model = BertClassificationFinetuner(
+                bert_model=model,
+                hidden_size=bert_config["hidden_size"],
+                num_classes=3 if FLAGS.task_name in ("mnli", "ax") else 2,
+            )
+            finetuning_model.compile(
+                optimizer=keras.optimizers.Adam(
+                    learning_rate=hp.Choice("lr", [5e-5, 4e-5, 3e-5, 2e-5])
+                ),
+                loss="sparse_categorical_crossentropy",
+                metrics=["accuracy"],
+            )
+            return finetuning_model
+
+    # Create a hypermodel object for a RandomSearch.
+    hypermodel = BertHyperModel()
+
+    # Initialize the random search over the 4 learning rate parameters, for 4
+    # trials and 3 epochs for each trial.
+    tuner = keras_tuner.RandomSearch(
+        hypermodel=hypermodel,
+        objective=keras_tuner.Objective("val_loss", direction="min"),
+        max_trials=4,
+        overwrite=True,
+        directory="tuning_hp_bert",
+        project_name="glue_finetuning_hp",
     )
-    finetuning_model.compile(
-        optimizer=keras.optimizers.Adam(learning_rate=FLAGS.learning_rate),
-        loss="sparse_categorical_crossentropy",
-        metrics=["accuracy"],
-    )
-    finetuning_model.fit(
-        train_ds, epochs=FLAGS.epochs, validation_data=validation_ds
+
+    tuner.search(train_ds, epochs=epochs, validation_data=validation_ds)
+
+    # Extract the best hyperparameters after the search.
+    best_hp = tuner.get_best_hyperparameters()[0]
+    finetuning_model = tuner.get_best_models()[0]
+
+    finetuning_model.fit(train_ds, epochs=epochs, validation_data=validation_ds)
```

Comment:
Is there a reason to retrain here? Does get_best_models not come with the trained weight during tuning?

Reply:
I think it returns an untrained model as in the [docs here](https://keras.io/api/keras_tuner/tuners/base_tuner/#getbestmodels-method), they say that it's recommended to retrain the model with the best hyperparameters for better performance.

Reply:
That appears to be saying you should use the best hyperparameters from get_best_hyperparameters and use that to retrain the model instead of get_best model. But we don't actually use get_best_hyperparameters for training, we just print out from there in the current version I think.

Reply:
[They say here](https://keras.io/api/keras_tuner/tuners/base_tuner/#getbesthyperparameters-method) that we should use the `get_best_hyperparameters` method to reinstantiate the `untrained best model` extracted from the `get_best_models` method I suppose.

Reply:
I don't think get_best_models will be untrained though? From the docs

> The models are loaded with the weights corresponding to their best checkpoint (at the end of the best epoch of best trial).

I think this current code is taking the *trained* best model and retraining it, which will double the number of epochs by the time we get to evaluation and saving! This is definitely incorrect, and could result in an overfit in these data light GLUE tasks.

I think if you want to do the retraining approach, you should call get_best_hyperparametrs, use that to call `hypermodel.build(best_hp)` to get a new model, then train that model from scratch.

Reply:
Yeah, I agree. I kind of don't understand why they recommend to retrained the best trained model? [This is what I found on StackOverflow](https://stackoverflow.com/questions/71773461/why-keras-tuner-get-best-models-only-returns-untrained-model), but I can't find a definite answer for this.

Reply:
[Here](https://github.com/keras-team/keras-tuner/blob/master/keras_tuner/engine/tuner.py#L345), it's mentioned that the trained models are returned from best to worse, and they recommend to retrain the model on the full dataset. So, yeah I suppose we can use `hypermodel.build()` itself, instead of retraining, just like you mentioned.
> Returns the best model(s), as determined by the objective.
        This method is for querying the models trained during the search.
        For best performance, it is recommended to retrain your Model on the
        full dataset using the best hyperparameters found during `search`,
        which can be obtained using `tuner.get_best_hyperparameters()

Reply:
We recommend retraining mainly because some users would use less epochs or less data during search to save time.
If every trial in the search fully trains the model, we would recommend to just use `get_best_models`.